File: android/src/main/java/com/mcodex/nitroarchive/infrastructure/storage/SafeDirectoryOutput.kt
package com.mcodex.nitroarchive.infrastructure.storage

import com.mcodex.nitroarchive.application.ports.DirectoryOutput
import com.mcodex.nitroarchive.domain.ArchiveDomainError
import java.io.File

class SafeDirectoryOutput(private val baseDir: File) : DirectoryOutput {
    private val createdPaths = mutableListOf<String>()

    private fun validatePath(path: String) {
        val resolved = File(baseDir, path).canonicalFile
        if (!resolved.path.startsWith(baseDir.canonicalPath)) {
            throw ArchiveDomainError.PathTraversal(path)
        }
    }

    override suspend fun prepareDirectory(at: String) {
        validatePath(at)
        val dir = File(baseDir, at)
        dir.mkdirs()
        createdPaths.add(at)
    }

    override suspend fun createFile(at: String): File {
        validatePath(at)
        val file = File(baseDir, at)
        file.parentFile?.mkdirs()
        file.createNewFile()
        createdPaths.add(at)
        return file
    }

    override suspend fun commit() {
        // For local directory output, commit is a no-op
    }

    override suspend fun rollback() {
        createdPaths.reversed().forEach { path ->
            File(baseDir, path).deleteRecursively()
        }
        createdPaths.clear()
    }
}

File: ios/Infrastructure/Storage/SafeDirectoryOutput.swift
import Foundation

final class SafeDirectoryOutput: DirectoryOutput {
  private let baseURL: URL
  private var createdPaths: [String] = []

  init(baseURL: URL) {
    self.baseURL = baseURL
  }

  func prepareDirectory(at path: String) async throws {
    let dirURL = baseURL.appendingPathComponent(path)
    guard dirURL.path.hasPrefix(baseURL.path) else {
      throw ArchiveDomainError.pathTraversal(path)
    }
    try FileManager.default.createDirectory(at: dirURL, withIntermediateDirectories: true, attributes: nil)
    createdPaths.append(path)
  }

  func createFile(at path: String) async throws -> URL {
    let fileURL = baseURL.appendingPathComponent(path)
    guard fileURL.path.hasPrefix(baseURL.path) else {
      throw ArchiveDomainError.pathTraversal(path)
    }
    let parent = fileURL.deletingLastPathComponent()
    try FileManager.default.createDirectory(at: parent, withIntermediateDirectories: true, attributes: nil)
    FileManager.default.createFile(atPath: fileURL.path, contents: nil)
    createdPaths.append(path)
    return fileURL
  }

  func commit() async throws {
    // For local directory output, commit is a no-op
  }

  func rollback() async throws {
    for path in createdPaths.reversed() {
      let url = baseURL.appendingPathComponent(path)
      try? FileManager.default.removeItem(at: url)
    }
    createdPaths.removeAll()
  }
}
